packages/components/src/components/Calculator/index.tsx

```diff
@@ -1,158 +1,277 @@
+  return Object.fromEntries(
+    calculator.inputs.map((input) => [input.name, defaultAsString(input)])
+  );
+}
 
 export const Calculator: FC<Props> = ({
-  value: calculator,
   environment,
   settings,
-  renderValue,
   valueWithContext,
 }) => {
-  const { path } = valueWithContext.context;
+  // useMemo here is important; valueWithContext.value is a getter that creates a new calculator value every time
+  const calculator = useMemo(() => valueWithContext.value, [valueWithContext]);
+
+  const inputNames = useMemo(
+    () => calculator.inputs.map((row) => row.name),
+    [calculator]
+  );
+
+  const path = useMemo(() => valueWithContext.context.path, [valueWithContext]);
+
   const { getLocalItemState, dispatch: viewerContextDispatch } =
     useViewerContext();
-  const itemState = getLocalItemState({ path });
-  const [prevCalculator, setPrevCalculator] = useState<SqCalculator | null>(
-    null
-  );
+  // Load just once on initial render, after the initial load, this component owns the state and pushes it to ViewerContext.
+  const [cachedState] = useState<CalculatorState | undefined>(() => {
+    const itemState = getLocalItemState({ path });
 
-  const getCalculatorStateFromCache = () => {
     const sameCalculatorCacheExists =
       itemState.calculator &&
-      hasSameCalculator(itemState.calculator, calculator);
+      itemState.calculator.hashString === calculator.hashString;
 
     if (sameCalculatorCacheExists) {
-      return itemState.calculator!;
+      return itemState.calculator;
     } else {
       return undefined;
     }
-  };
+  });
 
-  //It's possible that the calculator was changed when ths component was not visible. If that's the case, we want to reset it. We only want to use the cached version if its the same calculator.
-  const calculatorStateOnFirstRender = (calculator: SqCalculator) => {
-    const cache = getCalculatorStateFromCache();
-    return cache ? cache : initialCalculatorState(calculator);
-  };
+  const form = useForm({
+    defaultValues: cachedState?.codes ?? getFormValues(calculator),
+  });
 
-  const [calculatorState, calculatorDispatch] = useReducer(
-    adjustedReducer(path, viewerContextDispatch),
-    calculator,
-    calculatorStateOnFirstRender
+  const [inputValues, setInputValues] = useState<Record<string, SqValueResult>>(
+    () => cachedState?.inputValues ?? {}
   );
 
-  const _processAllFieldCodes = async () => {
-    await processAllFieldCodes({
-      dispatch: calculatorDispatch,
-      state: calculatorState,
-      path,
-      calculator,
-      environment,
-    });
-  };
+  // Takes all input codes and runs them all.  All values are updated simultaneously, to avoid too many rerenders.
+  const processAllFieldCodes = useCallback(async () => {
+    const codes = form.getValues();
+    const newInputValues: typeof inputValues = {};
+    for (const name of inputNames) {
+      const code = codes[name];
+      const valueResult = await runSquiggleCode(
```

Comment:
I feel like a lot of this is in-the-weeds for this file, can/should be isolated.